File: multithreading/fibonacci.py
import threading
import time

# Compute Fibonacci (CPU-intensive)
def fibonacci(n):
    if n <= 1:
        return n
    return fibonacci(n - 1) + fibonacci(n - 2)

# Worker function
def worker(n, results, thread_id):
    print(f"Thread {thread_id} computing Fibonacci({n})")
    results.append(fibonacci(n))

if __name__ == "__main__":
    num_threads = 4
    n = 35
    results = []
    threads = []

    start_time = time.time()

    # Start 4 threads
    for i in range(num_threads):
        thread = threading.Thread(target=worker, args=(n, results, i))
        threads.append(thread)
        thread.start()

    # Wait for all threads to complete
    for thread in threads:
        thread.join()

    for res in results:
        print("Result:", res)

    elapsed_time = time.time() - start_time
    print("Execution Time:", elapsed_time)

File: multithreading/fibonacci.go
package main

import (
	"fmt"
	"runtime"
	"sync"
	"time"
)

// Compute Fibonacci (CPU-intensive)
func fibonacci(n int) int {
	if n <= 1 {
		return n
	}
	return fibonacci(n-1) + fibonacci(n-2)
}

// Worker function
func worker(id int, n int, wg *sync.WaitGroup, results chan<- int) {
	defer wg.Done()
	fmt.Printf("Worker %d computing Fibonacci(%d)\n", id, n)
	results <- fibonacci(n)
}

func main() {
	runtime.GOMAXPROCS(runtime.NumCPU()) // Use all CPU cores

	numThreads := 4
	n := 35
	var wg sync.WaitGroup
	results := make(chan int, numThreads)

	start := time.Now()

	// Start 4 goroutines (threads)
	for i := 1; i <= numThreads; i++ {
		wg.Add(1)
		go worker(i, n, &wg, results)
	}

	wg.Wait()
	close(results)

	// Collect results
	for res := range results {
		fmt.Println("Result:", res)
	}

	elapsed := time.Since(start)
	fmt.Println("Execution Time:", elapsed)
}
